packages/shared/hooks/useAsync.ts

```diff
@@ -58,47 +56,106 @@ import { useCallback, useState } from 'react';
  *       return <UserAvatar url={fetchUserProfileAttempt.data.avatarUrl} />;
  *    }
  * }
+ *
+ * @example Aborting and useEffect cleanup function.
+ * If the first argument passed to the run function is an abort signal and the signal gets aborted,
+ * useAsync will ignore the return value of the promise from the callback.
+ *
+ * This lets you write a cleanup function for useEffect:
+ *
+ * const [attempt, run] = useAsync((message: string) => doAsyncStuff(message));
+ *
+ * useEffect(() => {
+ *   const abortController = new AbortController();
+ *   run(abortController.signal, 'hello');
+ *
+ *   return () => {
+ *     abortController.abort();
+ *   };
+ * });
  */
-export function useAsync<R, T extends Function>(cb?: AsyncCb<R, T>) {
-  const [state, setState] = useState<Attempt<R>>(makeEmptyAttempt);
+export function useAsync<Callback extends (...args: unknown[]) => unknown>(
+  cb?: Callback
+) {
+  const [state, setState] =
+    useState<Attempt<Awaited<ReturnType<Callback>>>>(makeEmptyAttempt);
 
   const run = useCallback(
-    (...p: Parameters<AsyncCb<R, T>>) =>
-      Promise.resolve()
+    (
+      ...args: Parameters<Callback> | [AbortSignal, ...Parameters<Callback>]
```

Comment:
So this means that TypeScript simply matches the args provided to `run` against two signatures: either you supply an array where the first element is an abort signal or you supply an array that's equal to the params accepted by the callback.

This leads to error messages which are not that great to be honest.

For example, if I have a callback which accepts no args but pass an extra string to it, this is what I get:

```
packages/shared/hooks/useAsync.test.ts:103:7 - error TS2345: Argument of type '[string]' is not assignable to parameter of type '[] | [AbortSignal]'.
  Type '[string]' is not assignable to type '[AbortSignal]'.
    Type 'string' is not assignable to type 'AbortSignal'.

103   run('foo');
```

The worst error message happens when you specify an extra arg to call where the first arg is not an abort signal. Let's say I have a callback which accepts two args and I supply an extra one:

```
packages/shared/hooks/useAsync.test.ts:179:7 - error TS2345: Argument of type '["foo", 2, "bar"]' is not assignable to parameter of type '[AbortSignal, string, number] | [message: string, count: number]'.
  Type '["foo", 2, "bar"]' is not assignable to type '[AbortSignal, string, number]'.
    Type at position 0 in source is not compatible with type at position 0 in target.
      Type 'string' is not assignable to type 'AbortSignal'.

179   run('foo', 2, 'bar');
```

This all stems from that simple matching that TypeScript performs.

An approach which would result in better error messages would be to add a generic type for the first arg of `run` and destructure the args within the function definition, as in `(firstArg, ...args)`.

Then we could tell TypeScript more or less this: if the first arg is an abort signal, assume that args is the params of the callback. If it's not an abort signal, then it must be the first param of the callback and args must be the rest of them.

I even managed to write such type:

```typescript
  const run = useCallback(
    <FirstArg extends AbortSignal | Parameters<RunFunc>[0]>(
      firstArg?: FirstArg,
      ...restOfArgs: FirstArg extends AbortSignal
        ? Parameters<RunFunc>
        : Parameters<RunFunc> extends [infer _, ...infer Tail]
        ? Tail
        : never
    ) => {
```

However, if you plug it in and look at the type errors in tests, you can see that it doesn't raise an error if the callback accepts one arg and you execute `run` with no args.
This is because the first arg is essentially treated as optional.

I didn't have time to figure out how to make this better, so for now I opted for the simpler solution.